### ISSUE TITLE
Change adaptive allocator size table to work on platforms with 32-bit Int

### ISSUE THREAD

**Issue by Lukasa** [apple/swift-nio#1848]
`AdaptiveRecvByteBufferAllocator.sizeTable` is calculated as follows:

https://github.com/apple/swift-nio/blob/c51907f551c6b2e4366b5ccbe8c0509b0b061026/Sources/NIO/RecvByteBufferAllocator.swift#L55-L71

On platforms where `Int` is 32-bits (such as watchOS), the loop starting on line 65 will either trap, as it attempts to stuff a `UInt32.max` into an `Int`, or it will trap, when we shift a 1 into the sign bit and attempt to coerce a negative `Int` into a `UInt32`. Either way, the loop does not function as intended.

**Comment by hassila** [comment]
Ouch. The sizetable should reasonably also be UInt, not much use to have negative alloc sizes.

**Comment by Lukasa** [comment]
While the size table can be `UInt`, the API of the type is specified in `Int`, so we will need to do transformations into `Int` _somewhere_ (going all the way down the stack we'll eventually get to `malloc`, which also takes `Int`). The better design is likely to keep this all as `Int` but not bother creating an alloc size any larger than `UInt32.max >> 1`, because frankly does anyone need an allocation block larger than 2GB?

**Comment by hassila** [comment]
Fair enough. Frankly, I'd probably never want a max Malloc block larger than a significantly smaller size in reality (say 256MB or so), the apps behavior is likely quite degenerate if it ends up with such large sizes anyway.
